src/librustc_mir/monomorphize/polymorphize.rs

```diff
@@ -0,0 +1,283 @@
+        // Can safely call `type_param` as all non-type parameters are considered used.
+        let generic_param_def = generics.type_param(&param, tcx);
+        debug!(
+            "emit_used_generic_parameters_error: generic_param_def={:?} param={:?}",
+            generic_param_def, param
+        );
+        let def_span = tcx.def_span(generic_param_def.def_id);
+        err.span_label(
+            def_span,
+            &format!("generic parameter `{}` is unused", generic_param_def.name),
+        );
+    }
+
+    err.emit();
+}
+
+/// Visitor used to aggregate generic parameter uses.
+struct UsedGenericParametersVisitor<'a, 'tcx> {
+    tcx: TyCtxt<'tcx>,
+    def_id: DefId,
+    body: ReadOnlyBodyAndCache<'a, 'tcx>,
+    used_parameters: &'a mut BitSet<u32>,
+}
+
+impl<'a, 'tcx> UsedGenericParametersVisitor<'a, 'tcx> {
+    fn visit_foldable<T: std::fmt::Debug + TypeFoldable<'tcx>>(&mut self, ty: T) {
+        debug!("visit_foldable: ty={:?}", ty);
+        // Exit early if this type doesn't use any type parameters.
+        if !ty.has_param_types() {
+            return;
+        }
+
+        // Use `TypeVisitor` to find any type parameters.
+        ty.visit_with(self);
+    }
+}
+
+impl<'a, 'tcx> Visitor<'tcx> for UsedGenericParametersVisitor<'a, 'tcx> {
+    fn visit_local_decl(&mut self, local: Local, local_decl: &LocalDecl<'tcx>) {
+        debug!("visit_local_decl: local_decl={:?}", local_decl);
+        let ty = local_decl
+            .ty
+            .builtin_deref(false)
+            .map(|type_and_mut| type_and_mut.ty)
+            .unwrap_or(local_decl.ty);
+        debug!("visit_local_decl: ty={:?}", ty);
+        // Skip visiting the closure that is currently being processed. This only happens
+        // because the first argument to the closure is a reference to itself and that
+        // will call `visit_substs`, resulting in each generic parameter captured being
+        // considered used by default.
+        match ty.kind {
+            ty::Closure(closure_def_id, _) if closure_def_id == self.def_id => {
+                debug!("visit_local_decl: skipping closure substs");
+                return;
+            }
+            _ => {}
+        }
+
+        self.super_local_decl(local, local_decl);
+    }
+
+    fn visit_const(&mut self, c: &&'tcx ty::Const<'tcx>, _: Location) {
+        debug!("visit_const: c={:?}", c);
+        // Necessary to detect `T::ASSOC_CONST`.
+        self.visit_foldable(*c);
+    }
+
+    fn visit_rvalue(&mut self, rvalue: &Rvalue<'tcx>, location: Location) {
+        debug!("visit_rvalue: rvalue={:?}", rvalue);
+        if let Rvalue::Cast(CastKind::Pointer(PointerCast::Unsize), ..) = *rvalue {
+            // If a closure performs dynamic dispatch then all of its generic parameters are
+            // used. This shouldn't be necessary but until there are changes to vtable
+            // infrastructure, it is.
+            self.used_parameters.insert_all();
```

Comment:
You don't need `insert_all` but rather to visit the source and target types in a mode that is more exhaustive, in that it doesn't special-case closures.

But also, I think the problem isn't related to this cast being in a closure, but rather a closure type being somewhere in the coerced type, which if we ignore its unused generics, would result in vtable polymorphization, which this PR doesn't implement.

Reply:
I've fixed this in 72e914d.